### DISCUSSION TITLE
How are container pid limits chosen for --pids-limit 0?

### DISCUSSION THREAD

**Discussion by jjenne-cisco** [containers/podman#15596] [Q&A]
I expect that the container's `pids.max` should be set to 'max', but this is only true for Cgroups v2. No for Cgroups v1.

Here are some datapoints on different VMs with cgroups v1 (ubuntu 20.04, podman 3.4.2) and cgroups v2 (ubuntu 22.04, podman 3.4.4):

```sh
root@cgroupsv1:~$ docker run ubuntu cat /sys/fs/cgroup/pids/pids.max
max
root@cgroupsv1:~$ podman run ubuntu cat /sys/fs/cgroup/pids/pids.max
2048
root@cgroupsv1:~$ podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max
7087
root@cgroupsv1:~$ podman run --pids-limit 0 --runtime /usr/sbin/runc ubuntu cat /sys/fs/cgroup/pids/pids.max
7087

root@cgroupsv2:~$ podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids.max
max
root@cgroupsv2:~$ podman run --pids-limit 0 --runtime /usr/sbin/runc ubuntu cat /sys/fs/cgroup/pids.max
7076
```

(Runtime comparisons for extra data points)

Firstly, I'm not exactly sure where the "unlimited" numbers are taken from as on my VMs `/sys/fs/cgroup/{machine,system,user}.slice/pids.max` are set to 'max'. Nowhere in the cgroup v1 hierarchy should podman be inheriting the above numbers.

I know that the default pid limit, when unlimited is not set on the CLI, is 2048 as set [here](https://github.com/containers/common/blob/f331dd84e10a5f6aee25b722f43ea9f29d274eb8/pkg/config/default.go#L149) and I can see that the `pids.max` file is being written with the limit [here](https://github.com/containers/podman/blob/2ad1fd3555de12de34e20898cc2ef901f08fe5ed/pkg/cgroups/pids.go#L32). But I can't work out the root of where the "unlimited" values are taken from and why `podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max` is not 'max' like docker's default.

Does anyone know any more about how or where the `resources.Pids.Limit` gets set in the case of `--pids-limit 0`?

---

EDIT: There is [an e2e test](https://github.com/containers/podman/blob/0085fbb488eec30e71e6cced6a06dbdb134e32a6/test/e2e/run_test.go#L1623) with the line "pids-limit not supported on cgroup V1" - I'm assuming this is an incorrect skip reason as `--pids-limit` seems in every way supported on cgroups v1.

**Comment by mheon** [comment]
@giuseppe PTAL

**Comment by giuseppe** [comment]
if you run rootless, the cgroups are not configured for the container, so it will inherit the cgroup from Podman.

**Reply by jjenne-cisco** [reply]
Sorry, I should mention that the above commands are all running as root. I'm interested in what the behaviour is for non-rootless (root-full?) podman.

**Reply by rhatdan** [reply]
We call it `rootful`.  In both cases the default is to inherit from the user process running the podman command.  If you don't tell Podman to set a pids-limit then the pids-limit of the podman process is inherited for the container.  Usually with root there is no Pids-limit.

**Reply by jjenne-cisco** [reply]
In these cases I've given though I _am_ telling Podman to set a pid limit. I would like the limit it to be unlimited/maximum. Like `docker`, does it not make sense for a user entering pids-limit unlimited to have "max" in `pids.max`?

It seems unusual that I can have a higher `pids.max` using a finite number than I can using "unlimited".
```
root@ubuntu:~# podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max
7093  # as you say, inherited from some parent process cgroup
root@ubuntu:~# podman run --pids-limit 10000 ubuntu cat /sys/fs/cgroup/pids/pids.max
10000
```

As a side, not as important, note: I also can't see any process/cgroup that podman is inheriting from with the same pid limit, so I'm also not sure how it is choosing 7093:
```
root@ubuntu:~# find /sys/fs/cgroup -name pids.max | xargs grep 7093
root@ubuntu:~# podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max
7093
root@ubuntu:~#
```
But I can see when the ctr cgroup is _not_ created that the root cgroup `pids.max` is picked up.
```
root@ubuntu:~# podman run --cgroups=disabled ubuntu cat /sys/fs/cgroup/pids/pids.max
max
```

**Reply by jjenne-cisco** [reply]
As an additional datapoint, it looks like setting the Cgroup manager to `cgroupfs` sets the right, expected pid limit. It looks to be an issue with the `systemd` manager if that makes sense?

```
root@ubuntu:~# podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max
7093
root@ubuntu:~# podman run --pids-limit 0 --cgroup-manager=cgroupfs ubuntu cat /sys/fs/cgroup/pids/pids.max
max
```

**Reply by rhatdan** [reply]
Does this only work on cgroupv1?

**Reply by jjenne-cisco** [reply]
Yes, the datapoints above are all on a cgroupsv1 host. On cgroupsv2 I see "max" like I'd expect:
```
root@cgroupsv2:~$ podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids.max
max
```

**Reply by jjenne-cisco** [reply]
On trying docker + systemd cgroup driver, it looks like this is related to cgroupsv1 and systemd cgroup manager:
```
root@ubuntu:~# cat /etc/docker/daemon.json
{
  "ipv6": true,
  "fixed-cidr-v6": "fd00::/80"
}
root@ubuntu:~# docker run ubuntu cat /sys/fs/cgroup/pids/pids.max
max
root@ubuntu:~# vim /etc/docker/daemon.json
root@ubuntu:~# cat /etc/docker/daemon.json
{
  "ipv6": true,
  "fixed-cidr-v6": "fd00::/80",
  "exec-opts": ["native.cgroupdriver=systemd"]
}
root@ubuntu:~# systemctl restart docker
root@ubuntu:~# docker run ubuntu cat /sys/fs/cgroup/pids/pids.max
7093
```

**Reply by rhatdan** [reply]
Which probably means that the issue is with systemd, On cgroup v1, so not likely to be fixed.

**Reply by jjenne-cisco** [reply]
Yes, thanks, you're right
```
root@ubuntu:~# podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max
7093
root@ubuntu:~# echo 'DefaultTasksMax=infinity' >> /etc/systemd/system.conf
root@ubuntu:~# systemctl daemon-reexec
root@ubuntu:~# podman run --pids-limit 0 ubuntu cat /sys/fs/cgroup/pids/pids.max
max
root@ubuntu:~# podman run ubuntu cat /sys/fs/cgroup/pids/pids.max
2048
```
As you say, systemd assigns the pids/tasks max when it creates the container cgroup. Therefore in cgroups v1 we get "unlimited pids" _only_ when podman is passed the "unlimited pids" argument **_and also_** systemd config is added to set "unlimited pids" by default.

I think what has been confusing is that while the help output for `--pids-limit 0` says "for unlimited", in reality what it means is "if specified podman will not overwrite whatever value has been set by the cgroup driver".

**Comment by legaul-cisco** [comment]
Hey all :)

(disclaimer: I work with @jjenne-cisco and share the interest in this behaviour)

Some of my thoughts around the above (and referring to https://docs.podman.io/en/latest/markdown/options/pids-limit.html):
- Podman has a default pid limit
  - The docs say "The default is 4096 on systems that support “pids” cgroup controller."
  - Was the default limit changed from 2048 to 4096 at some point?
  - This is one of the few ways podman's behaviour deviates from docker, which has no default limit
  - I'm not sure why there's a default pid limit when there's no defaults for other limits such as memory - how is podman supposed to know what a sensible default is for all containers?
- The default pids limit can be removed by passing `--pids-limit -1` or `--pids-limit 0`
  - The podman docs say "Set to -1 to have unlimited pids for the container."
  - I think `--pids-limit 0` is allowed for podman backwards compatibility maybe?
  - Docker similarly supports `--pids-limit` with "Tune container pids limit (set -1 for unlimited)"
- Despite what the docs say, `--pids-limit -1` isn't actually allowing unlimited pids in the container
  - When using the systemd cgroup driver there is a limit being set by systemd
  - systemd as the cgroup driver is the default with podman, but not docker on cgroups v1?
  - I'm guessing `--pids-limit -1` is just being interpreted by podman as "don't set anything in `pids.max`", rather than "set `max` in `pids.max`"?
  - `--pids-limit max` is rejected by podman
- The behaviour on cgroups v2 seems to be different, i.e. no default limit being set by systemd (unsure why)

My vote/interpretation would be:
- Podman should have no default pid limit (systemd may set the value, as seen here)
- `--pids-limit -1` (or perhaps `--pids-limit max`?) should explicitly set the limit to 'max' (overwriting whatever systemd might set)
- Question over whether this pid limit should be being set by systemd, although I suspect this is intended behaviour

As things stand, the best we can do is pass `--pids-limit <N>` where `<N>` is some arbitrarily large pid limit... It seems strange that this would be the recommended solution rather than an argument documented to set the limit to unlimited ('max').

Any thoughts welcome :)

**Reply by rhatdan** [reply]
Code says the default should be 2048
DefaultPidsLimit = 2048

**Comment by rhatdan** [comment]
On Fedora 36 with cgroups v2, I see:

```
$ podman run alpine cat /sys/fs/cgroup/pids.max
2048
$ podman run --pids-limit -1 alpine cat /sys/fs/cgroup/pids.max
max
$ podman run --pids-limit 0 alpine cat /sys/fs/cgroup/pids.max
max
```

**Reply by legaul-cisco** [reply]
Yeah, agreed this isn't a problem on cgroups v2 (not exactly sure why there's a difference). Podman still supports cgroups v1 though right?

**Reply by rhatdan** [reply]
Yes
